mlir/lib/Dialect/Tosa/Transforms/SinkInputOpsThroughConcat.cpp

```diff
@@ -0,0 +1,425 @@
+
+namespace {
+struct SinkGenericOp : public OpRewritePattern<tosa::ConcatOp> {
+  SinkGenericOp(MLIRContext *context, PatternBenefit benefit,
+                llvm::StringMap<unsigned> &stats, llvm::raw_ostream &os)
+      : OpRewritePattern<tosa::ConcatOp>::OpRewritePattern(context, benefit),
+        stats(stats), os(os) {}
+
+  LogicalResult matchAndRewrite(tosa::ConcatOp concatOp,
+                                PatternRewriter &rewriter) const override {
+
+    Operation *sample = nullptr;
+    for (auto val : concatOp->getOperands()) {
+      if (!val.hasOneUse())
+        return rewriter.notifyMatchFailure(
+            concatOp, "Operands must just connect to this concat.");
+
+      auto *genericOp = val.getDefiningOp();
+      if (!genericOp)
+        return rewriter.notifyMatchFailure(
+            concatOp, "Requires all operands to be operators");
+
+      if (!sample)
+        sample = genericOp;
+
+      if (sample->getName().getIdentifier() !=
+          genericOp->getName().getIdentifier())
+        return rewriter.notifyMatchFailure(
+            concatOp, "Requires all operands to be the same");
+    }
+    auto opName = sample->getName().getStringRef();
+    auto amount = ++stats[opName];
+    if (amount == 1 || amount % 50 == 0)
+      os << "SinkInputOpsThroughConcat: Operation " << opName
+         << " -> Matched amount: " << amount << "\n";
+    return rewriter.notifyMatchFailure(concatOp,
+                                       "Only for statistics printing");
+  }
+
+private:
+  llvm::StringMap<unsigned> &stats;
+  llvm::raw_ostream &os;
+};
+
+struct CheckResult {
+  bool rewriteIsPossible;
+  const char *msg;
+};
+
+template <class OpT>
+struct GenericCheckResult {
+  OpT operand;
+  const char *msg;
+};
+
+template <class OpT>
+bool checkForEquality(OpT sample, OpT op) {
+  for (auto attrName : OpT::getAttributeNames()) {
+    if (sample->getAttr(attrName) != op->getAttr(attrName)) {
+      return false;
+    }
+  }
+  return llvm::equal(op->getOperandTypes(), sample->getOperandTypes());
+}
+
+template <class OpT>
+struct SinkSpecificOp : public OpRewritePattern<tosa::ConcatOp> {
+  using OpRewritePattern<tosa::ConcatOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tosa::ConcatOp concatOp,
+                                PatternRewriter &rewriter) const override {
+    auto genChecks = doGenericChecks(concatOp);
+    auto sample = genChecks.operand;
+    if (!sample) {
+      return rewriter.notifyMatchFailure(concatOp, genChecks.msg);
+    }
+
+    const auto extraChecks = this->doExtraChecks(sample, concatOp);
+    if (!extraChecks.rewriteIsPossible) {
+      return rewriter.notifyMatchFailure(concatOp, extraChecks.msg);
+    }
+
+    // rewriting, first collect all inputs and rearrange them for the concats
+    SmallVector<SmallVector<Value>> concatOperands(sample->getNumOperands());
+    for (auto val : concatOp.getOperands()) {
+      auto *rewriteOp = val.getDefiningOp();
```

Comment:
```suggestion
      auto *producerOp= val.getDefiningOp();
```
This op is not getting rewritten?